### PR TITLE
feat(doctor): expand provider preflight with auth / env / timeout / CLI display (TRN-3021)

### LINKED ISSUE #38: TRN-3021: Expand provider doctor preflight checks
## One-paragraph summary
`trinity doctor` already validates provider configuration and command availability, but several failures still surface only when review execution starts, such as missing login, unsupported model options, or weak timeout configuration.

## Scope hints (optional)
- Add clearer provider preflight diagnostics where feasible without making expensive model calls.
- Check common CLI presence and executable permissions.
- Consider lightweight version or auth checks for supported providers.
- Report model, reasoning effort, timeout, and optional-provider status clearly.

## Out of scope
- Running real reviews from doctor.
- Requiring network calls for every provider.
- Failing optional providers unless explicitly requested.

## Verification
- Fake CLI tests for success and failure modes.
- Doctor output snapshot or stable assertion tests.
- `make test`, `make lint`, and `af validate --root .`.

### PR DESCRIPTION
**Pull request by ryosaeba1985**
## Summary

`trinity doctor` now surfaces 4 classes of configuration issue that previously only showed up at review-execution time:

1. **Wrapper-provider auth state** (deepseek + openrouter) — env-var-or-file precedence with mode 600/400 check, mirroring `providers/bin/<wrapper>:10-31` exactly
2. **Timeout sanity** — warns when configured timeout < 60s
3. **Shell env pollution** — lists vars matching the TRN-3023 spawn-time clearlist (`*_BASE_URL`, `OTEL_*`, etc.) so operators can audit `direnv` / shell setups
4. **Resolved CLI string** per provider (shlex-joined for safe quoting)

Plus REQUIRED vs OPTIONAL split driven by the active preset's metadata. Required-provider issues are fatal (exit 1); optional-provider issues are demoted to warnings (exit 0).

Closes #38.

## What changed

- **`scripts/codex.py`**:
  - 3 new module constants: `_MIN_TIMEOUT_WARNING_SECONDS = 60`, `_WRAPPER_AUTH_CONFIG`, plus reuse of TRN-3023's clearlist constants
  - 5 new helpers: `_make_health_result`, `wrapper_auth_check`, `detect_env_pollution`, `_format_provider_block`, plus extension to `provider_health` / `provider_health_results` / `format_health_results` / `health_results_ok`
  - `resolve_preset_providers` extended with 2 NEW metadata keys (`providers`, `optional_providers`); existing 5 keys preserved verbatim
  - `cmd_doctor` threads `preset_metadata` to BOTH the formatter AND `health_results_ok` exit-code decision
- **`tests/test_doctor_preflight.py` (NEW)**: 34 unit tests covering A1-A17 + edge cases (mode boundary 600/400/644, env-precedence, env-pollution sort + redaction + essentials, REQUIRED/OPTIONAL fatality demotion, format backwards-compat, end-to-end wrong-mode → REQUIRED FAIL chain)
- **`tests/test_codex_adapter.py`**: 1 metadata-shape fixture updated; 1 doctor test now sets `DEEPSEEK_API_KEY` (since post-TRN-3021 doctor checks auth)
- **`README.md`**: Provider health section expanded
- **`CHANGELOG.md`**: `[Unreleased]` entries under both `### Added` (new behavior) and `### Changed` (`format_health_results` parameterization for cmd_review co-consumer compatibility)

## Backwards compatibility

`cmd_review --check-providers` continues to use the existing single-line format (`{provider}: OK - {executable} (timeout Ns)`) — `format_health_results` defaults to `verbose=False` and the call sites at codex.py:1664-1681 don't pass any new kwargs. `health_results_ok` defaults `preset_metadata=None` which treats all providers as REQUIRED, preserving existing exit-code semantics. First line per provider stays grep-compatible even in verbose mode.

## Trinity panel approval

Both gates met (all 4 providers individually ≥9.0):

| Phase | Rounds | gemini | codex | glm | deepseek | Mean |
|-------|--------|--------|-------|-----|----------|------|
| Plan-review | 6 | 9.2 (R2) | 9.1 (R6) | 9.0 (R3) | 9.0 (R4) | **9.075** |
| Code-review | 1 | 9.8 | 9.1 | 9.4 | 9.4 | **9.43** |

**Plan-review took 6 rounds** because codex was the toughest reviewer. Each round caught real architectural defects before code was written:
- R1: registry plumbing gap (codex.py has zero refs to registry.json) + shared formatter co-consumer + warning schema unsafe + synthetic-error branch coverage
- R2: REQUIRED/OPTIONAL fatality wiring + preset_metadata schema mismatch  
- R3: cmd_doctor return-path wiring + auth_files_status name drift + `requested` semantics regression
- R4: wrong-mode auth fatality alignment + helper name drift
- R5: line 98 unconditional warnings list + line 86 missing-auth severity routing
- R6: PASS

**Methodology lesson logged**: deepseek correctly flagged TRN-1007 phantom reference across R1/R2/R3 — my branch was created off stale local main pre-PR-#67-merge so TRN-1007 wasn't on the branch. The "false positive" dismissal was wrong. Future CHGs should `git status -uno` against `origin/main` before R1 dispatch.

## Acceptance (per TRN-1007 — first CHG to dogfood the new SOP)

- [x] **§1 README**: Provider health section expanded
- [x] **§1 CHANGELOG**: `[Unreleased]` gained `### Added` + `### Changed` entries naming TRN-3021 behavior + format change
- [x] **§1 SKILL/providers**: ➖ N/A — no provider behavior change
- [x] **§2 verification**: `make test` (218 + 34 = 252 passed), `make lint` (clean), `make coverage` 84%, `af validate` (114 / 0 issues), `make verify-built` clean
- [x] **§3 drift**: ➖ N/A — no provider CLI changes
- [x] **§4 6-step methodology**: applied throughout 6 plan-review rounds + code-review; each finding resolved with line citations
- [x] **§5 manual smoke**: `OPENAI_BASE_URL=https://example.invalid` set in shell → `detect_env_pollution()` flags it, value redacted to first 12 chars + …
- [x] **§6 identity**: `gh auth status` → `ryosaeba1985`
- [x] **§7 branch hygiene**: rebased on origin/main; branch `codex/trn-3021-doctor-preflight`

## Test plan

- [ ] CI ubuntu-latest passes (`make test` clean, `make coverage` ≥80%)
- [ ] CI macos-latest passes
- [ ] Codex GitHub bot review

Closes #38

### CODE REVIEW

**Review by frankyxhl — DISMISSED** [review]

**Comment by frankyxhl** [comment]
@codex review

File: scripts/codex.py [hunk 1]
```diff
@@ -403,70 +552,199 @@ def provider_health(provider, provider_config, root):
...
-    }
+    if timeout is not None and timeout < _MIN_TIMEOUT_WARNING_SECONDS:
+        warnings.append(
+            f"timeout {timeout}s is below {_MIN_TIMEOUT_WARNING_SECONDS}s "
+            f"minimum recommended"
+        )
+        timeout_warning = True
+
+    auth = wrapper_auth_check(provider)
+    if auth is not None:
```
[the remark this answers is not in the document]
Comment (ryosaeba1985) [reply scripts/codex.py]: This finding was on commit 9a342d3, before the CI fix at aa4c906 landed. The architectural fix (out of band with this bot review) addresses it: `wrapper_auth_check` is no longer called from `provider_health` (which is shared with `cmd_review --check-providers`). It now runs only in `cmd_doctor`, gated on `_is_canonical_wrapper(executable, provider)` which matches `/skills/trinity/bin/<provider>` substring or `/providers/bin/<provider>` suffix. Custom configs that reuse the provider names with a different CLI path will not match the gate, so doctor and review both work cleanly. Verified locally: 251 tests pass on aa4c906; bot's 12 originally-failing CI tests now green.

File: scripts/codex.py [hunk 2]
```diff
@@ -403,70 +580,188 @@ def provider_health(provider, provider_config, root):
...
+    OPTIONAL providers with `issues` non-empty → demoted (don't fail).
+    `preset_metadata=None` (existing call sites in cmd_review --check-providers)
+    treats all providers as REQUIRED, preserving current semantics.
+    """
+    optional_set = set()
+    if preset_metadata is not None:
+        optional_set = set(preset_metadata.get("optional_providers", []))
+    for result in results:
+        if result.get("issues") and result["provider"] not in optional_set:
+            return False
```
[the remark this answers is not in the document]
Comment (ryosaeba1985) [reply scripts/codex.py:798]: Fixed in afde482. Both `health_results_ok` and `format_health_results` now subtract the required set before treating the rest as optional:

\`\`\`python
required_set = set(preset_metadata.get("providers", []))
optional_set = set(preset_metadata.get("optional_providers", [])) - required_set
\`\`\`

Regression test \`test_a14_overlap_required_wins\` asserts: glm in BOTH `providers` and `optional_providers`, with non-empty `issues` → `health_results_ok` returns False (exit 1), not silently True.

File: scripts/codex.py [hunk 3]
```diff
@@ -370,6 +378,159 @@ def command_has_path(command):
...
+            "source": "file",
+            "env_var": env_name,
+            "file": str(file_path),
+            "mode_ok": False,
+        }
+    return {
+        "source": "file",
+        "env_var": env_name,
+        "file": str(file_path),
+        "mode_ok": mode in (0o600, 0o400),
```
[the remark this answers is not in the document]
Comment (ryosaeba1985) [reply scripts/codex.py:485]: Fixed in afde482. \`wrapper_auth_check\` now uses \`file_path.lstat()\` instead of \`stat()\` so the symlink's OWN mode is checked (matches wrapper's \`stat -c '%a' "\$KEY_FILE"\` behavior on GNU/Linux which reports the symlink mode, typically 777, not the target's).

Regression test \`test_a4c_symlink_uses_lstat\` creates a symlink to a 0600 target file, asserts \`mode_ok is False\` (because the symlink's own mode is 0o755 on macOS / 0o777 on Linux). Without the lstat fix, doctor would report OK while wrappers refused at runtime.

File: scripts/codex.py [hunk 4]
```diff
@@ -403,70 +586,199 @@ def provider_health(provider, provider_config, root):
...
-        if result["ok"]:
+    if result.get("ok"):
+        first = (
+            f"{provider}: OK - {result.get('executable')} "
+            f"(timeout {result.get('timeout')}s)"
+        )
+    else:
+        # First non-fatal warning OR first issue feeds the headline.
+        if result.get("issues"):
+            first = f"{provider}: FAIL - {result['issues'][0]}"
```
[the remark this answers is not in the document]
Comment (ryosaeba1985) [reply scripts/codex.py]: Fixed in dab54b8. Re-introduced `optional` keyword arg on `_format_provider_block`; when True the headline label is `WARN` (not `FAIL`):

\`\`\`python
def _format_provider_block(result, *, optional=False):
    ...
    fail_label = "WARN" if optional else "FAIL"
    if result.get("ok"):
        first = f"{provider}: OK - ..."
    else:
        first = f"{provider}: {fail_label} - {result['issues'][0]}"
\`\`\`

\`format_health_results\` passes \`optional=True\` only in the OPTIONAL section iteration; REQUIRED stays default. Display now matches \`health_results_ok\`'s exit-code semantics.

Regression test \`test_optional_provider_issues_render_as_warn_not_fail\` asserts: optional provider with non-empty \`issues\` → output contains \`codex: WARN -\` and does NOT contain \`codex: FAIL -\`. Verified locally: 254 tests pass.

File: scripts/codex.py [hunk 5]
```diff
@@ -370,6 +378,165 @@ def command_has_path(command):
...
+    that's the point: doctor's wrapper-auth check should only fire when
+    the user has actually installed the canonical wrapper.
+    """
+    if executable is None or provider not in _WRAPPER_AUTH_CONFIG:
+        return False
+    p = str(executable)
+    # Match the canonical install root pattern (claude or codex).
+    needle = f"/skills/trinity/bin/{provider}"
+    if needle in p:
+        return True
```
[the remark this answers is not in the document]
Comment (ryosaeba1985) [reply scripts/codex.py]: Fixed in the next push. \`_is_canonical_wrapper\` now compares \`Path(executable).resolve()\` against an exact list of canonical install paths (no more substring matching):

\`\`\`python
home = Path(os.path.expanduser("~"))
candidates = [
    home / ".claude" / "skills" / "trinity" / "bin" / provider,
    home / ".codex" / "skills" / "trinity" / "bin" / provider,
    Path(__file__).resolve().parent.parent / "providers" / "bin" / provider,
]
for c in candidates:
    if c.resolve() == actual:
        return True
return False
\`\`\`

Custom configs reusing the provider name with a different executable (`deepseek-test`) or path (`/tmp/skills/trinity/bin/deepseek`) won't match → cmd_doctor stays out of their way.

2 regression tests added: \`test_canonical_wrapper_rejects_substring_match\` (basename trap + non-canonical-root trap) and \`test_canonical_wrapper_resolves_symlinks\` (symlinked install still matches via resolve()). 256 tests pass locally.

**Review by frankyxhl — APPROVED** [review]
Approved by maintainer. Reviewed via local watchdog (SWM-1103):

- head: 8badd1be
- CI:
✓ macos-latest: SUCCESS
✓ ubuntu-latest: SUCCESS
- Codex bot: approved
- inline findings: 0 open, 6 resolved

Reason: head 8badd1be, CI green on ubuntu+macos, Codex bot 👍 approved, 6/6 threads resolved (Stage 1.5 synced 3 threads across the iteration), prior frankyxhl review auto-dismissed by branch protection on head bump
